2007-06-07  Dan Williams  <dcbw@redhat.com>

* (nm_device_802_3_ethernet_link_activated,
   nm_device_802_3_ethernet_link_deactivated): fix ethernet link detection
	behavior due to races between the netlink code and when the
	device thread recognizes the change.  Checking the link state
	before scheduling the change even on the device thread is wrong.
	(gnome.org #354565, rh #194124)

--

--- bzr.debian.0.9.5/src/nm-device-802-3-ethernet.c	(revision 2577)
+++ bzr.debian.0.9.5.new/src/nm-device-802-3-ethernet.c	(revision 2578)
@@ -108,18 +108,16 @@
                                          GObject *obj,
                                          NMDevice8023Ethernet *self)
 {
+	GSource * source;
+
 	/* Make sure signal is for us */
 	if (NM_DEVICE (self) != NM_DEVICE (obj))
 		return;
 
-	if (!nm_device_has_active_link (NM_DEVICE (self)))
-	{
-		GSource *	source = g_idle_source_new ();
-
-		g_source_set_callback (source, (GSourceFunc) link_activated_helper, self, NULL);
-		g_source_attach (source, nm_device_get_main_context (NM_DEVICE (self)));
-		g_source_unref (source);
-	}
+	source = g_idle_source_new ();
+	g_source_set_callback (source, (GSourceFunc) link_activated_helper, self, NULL);
+	g_source_attach (source, nm_device_get_main_context (NM_DEVICE (self)));
+	g_source_unref (source);
 }
 
 
@@ -135,18 +133,16 @@
                                            GObject *obj,
                                            NMDevice8023Ethernet *self)
 {
+	GSource * source;
+
 	/* Make sure signal is for us */
 	if (NM_DEVICE (self) != NM_DEVICE (obj))
 		return;
 
-	if (nm_device_has_active_link (NM_DEVICE (self)))
-	{
-		GSource *	source = g_idle_source_new ();
-
-		g_source_set_callback (source, (GSourceFunc) link_deactivated_helper, self, NULL);
-		g_source_attach (source, nm_device_get_main_context (NM_DEVICE (self)));
-		g_source_unref (source);
-	}
+	source = g_idle_source_new ();
+	g_source_set_callback (source, (GSourceFunc) link_deactivated_helper, self, NULL);
+	g_source_attach (source, nm_device_get_main_context (NM_DEVICE (self)));
+	g_source_unref (source);
 }
 
 static void
